x/oracle/keeper/keeper.go

```diff
@@ -183,6 +160,77 @@ func (k Keeper) SetFeeInfo(ctx sdk.Context, networkDescriptor types.NetworkDescr
 	return nil
 }
 
+// SetProphecyWithInitValue set the prophecy in keeper
+func (k Keeper) SetProphecyWithInitValue(ctx sdk.Context, prophecyID []byte) {
+	prophecy := types.Prophecy{
+		Id:              prophecyID,
+		Status:          types.StatusText_STATUS_TEXT_PENDING,
+		ClaimValidators: []string{},
+	}
+	k.SetProphecy(ctx, prophecy)
+}
+
+// ProcessSignProphecy deal with the signature from validator
+func (k Keeper) ProcessSignProphecy(ctx sdk.Context, networkDescriptor types.NetworkDescriptor, prophecyID []byte, cosmosSender, ethereumAddress, signature string) error {
+	prophecy, ok := k.GetProphecy(ctx, prophecyID)
+	if !ok {
+		return types.ErrProphecyNotFound
+	}
+
+	// verify the signature
+	publicKey, err := gethCrypto.Ecrecover(prophecyID, gethCommon.FromHex(signature))
+	if err != nil {
+		return err
+	}
+
+	ok = gethCrypto.VerifySignature(publicKey, prophecyID, []byte(signature))
+	if !ok {
+		return errors.New("incorrect signature")
+	}
+
+	valAddr, err := sdk.ValAddressFromBech32(cosmosSender)
+	if err != nil {
+		return err
+	}
+
+	err = k.AppendSignature(ctx, prophecyID, ethereumAddress, signature)
+	if err != nil {
+		return err
+	}
+	oldStatus := prophecy.Status
+
+	newStatus, err := k.AppendValidatorToProphecy(ctx, networkDescriptor, prophecyID, valAddr)
+
+	// emit the event when status from pending to success
+	if oldStatus == types.StatusText_STATUS_TEXT_PENDING && newStatus == types.StatusText_STATUS_TEXT_SUCCESS {
+		prophecyInfo, ok := k.GetProphecyInfo(ctx, prophecyID)
+		if !ok {
+			return errors.New("prophecy info not available in keeper")
+		}
+
+		ctx.EventManager().EmitEvents(sdk.Events{
+
+			sdk.NewEvent(
+				types.EventTypeProphecyCompleted,
+				sdk.NewAttribute(types.AttributeKeyProphecyID, string(prophecyID)),
+				sdk.NewAttribute(types.AttributeKeyNetworkDescriptor, prophecyInfo.NetworkDescriptor.String()),
+				sdk.NewAttribute(types.AttributeKeyCosmosSender, prophecyInfo.CosmosSender),
+				sdk.NewAttribute(types.AttributeKeyCosmosSenderSequence, strconv.FormatInt(int64(prophecyInfo.CosmosSenderSequence), 10)),
+				sdk.NewAttribute(types.AttributeKeyEthereumReceiver, prophecyInfo.EthereumReceiver),
+				// TODO get the token address from denom
+				sdk.NewAttribute(types.AttributeKeyTokenContractAddress, prophecyInfo.EthereumReceiver),
+				sdk.NewAttribute(types.AttributeKeyAmount, strconv.FormatInt(prophecyInfo.TokenAmount.Int64(), 10)),
+				sdk.NewAttribute(types.AttributeKeyDoublePeggy, strconv.FormatBool(prophecyInfo.DoublePeg)),
+				sdk.NewAttribute(types.AttributeKeyGlobalNonce, strconv.FormatInt(int64(prophecyInfo.GlobalNonce), 10)),
+				sdk.NewAttribute(types.AttributeKeycrossChainFee, strconv.FormatInt(prophecyInfo.CrosschainFee.Int64(), 10)),
+				sdk.NewAttribute(types.AttributeKeySignatures, strings.Join(prophecyInfo.Signatures, ",")),
+				sdk.NewAttribute(types.AttributeKeyEthereumAddresses, strings.Join(prophecyInfo.EthereumAddress, ",")),
```

Comment:
Why would we emit an event with an array of ethereum addresses? There can only be 1 receiver and 1 token address.

Reply:
when a prophecy become completed, the relayer need get all signatures and all addresses, then relayer can call the smart contract with all data needed.

Reply:
The ethereum address array is the array of the signers?